File: backend/app/Models/SharedRide.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class SharedRide extends Model
{
    use HasFactory;

    protected $fillable = [
        'driver_id',
        'vehicle_id',
        'pickup_latitude',
        'pickup_longitude',
        'pickup_address',
        'dropoff_latitude',
        'dropoff_longitude',
        'dropoff_address',
        'departure_time',
        'max_passengers',
        'current_passengers',
        'price_per_seat',
        'status',
    ];

    protected $casts = [
        'pickup_latitude' => 'float',
        'pickup_longitude' => 'float',
        'dropoff_latitude' => 'float',
        'dropoff_longitude' => 'float',
        'departure_time' => 'datetime',
        'price_per_seat' => 'decimal:2',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    protected $appends = ['has_available_seats', 'is_full'];

    /**
     * Get the driver of this shared ride.
     */
    public function driver()
    {
        return $this->belongsTo(User::class, 'driver_id');
    }

    /**
     * Get the vehicle used for this shared ride.
     */
    public function vehicle()
    {
        return $this->belongsTo(Vehicle::class);
    }

    /**
     * Get the passengers in this shared ride.
     */
    public function passengers()
    {
        return $this->hasMany(SharedRidePassenger::class);
    }

    /**
     * Get active passengers only.
     */
    public function activePassengers()
    {
        return $this->passengers()->whereIn('status', ['confirmed', 'picked_up']);
    }

    /**
     * Check if there are available seats.
     */
    public function getHasAvailableSeatsAttribute()
    {
        return $this->current_passengers < $this->max_passengers;
    }

    /**
     * Check if the ride is full.
     */
    public function getIsFullAttribute()
    {
        return $this->current_passengers >= $this->max_passengers;
    }

    /**
     * Scope to filter scheduled rides.
     */
    public function scopeScheduled($query)
    {
        return $query->where('status', 'scheduled');
    }

    /**
     * Scope to filter rides with available seats.
     */
    public function scopeWithAvailableSeats($query)
    {
        return $query->whereColumn('current_passengers', '<', 'max_passengers');
    }

    /**
     * Scope to search rides by location and time.
     */
    public function scopeSearchByRoute($query, $pickupLat, $pickupLng, $dropoffLat, $dropoffLng, $maxDistance = 5)
    {
        // Using Haversine formula to calculate distance
        $earthRadius = 6371; // km

        return $query->selectRaw("
            *,
            (
                $earthRadius * acos(
                    cos(radians(?)) * cos(radians(pickup_latitude)) *
                    cos(radians(pickup_longitude) - radians(?)) +
                    sin(radians(?)) * sin(radians(pickup_latitude))
                )
            ) as pickup_distance,
            (
                $earthRadius * acos(
                    cos(radians(?)) * cos(radians(dropoff_latitude)) *
                    cos(radians(dropoff_longitude) - radians(?)) +
                    sin(radians(?)) * sin(radians(dropoff_latitude))
                )
            ) as dropoff_distance
        ", [
            $pickupLat, $pickupLng, $pickupLat,
            $dropoffLat, $dropoffLng, $dropoffLat
        ])
        ->havingRaw('pickup_distance < ?', [$maxDistance])
        ->havingRaw('dropoff_distance < ?', [$maxDistance])
        ->orderBy('pickup_distance');
    }

    /**
     * Update passenger count.
     */
    public function updatePassengerCount()
    {
        $this->current_passengers = $this->activePassengers()->count();
        $this->save();
    }
}

File: packages/mobi_core/lib/src/models/shared_ride.dart
import 'package:equatable/equatable.dart';

/// Shared ride model for carpooling functionality
class SharedRide extends Equatable {
  final int id;
  final int driverId;
  final int? vehicleId;
  final double pickupLatitude;
  final double pickupLongitude;
  final String pickupAddress;
  final double dropoffLatitude;
  final double dropoffLongitude;
  final String dropoffAddress;
  final DateTime departureTime;
  final int maxPassengers;
  final int currentPassengers;
  final double pricePerSeat;
  final String status; // 'scheduled', 'in_progress', 'completed', 'cancelled'
  final List<SharedRidePassenger> passengers;
  final DateTime createdAt;
  final DateTime updatedAt;

  const SharedRide({
    required this.id,
    required this.driverId,
    this.vehicleId,
    required this.pickupLatitude,
    required this.pickupLongitude,
    required this.pickupAddress,
    required this.dropoffLatitude,
    required this.dropoffLongitude,
    required this.dropoffAddress,
    required this.departureTime,
    required this.maxPassengers,
    required this.currentPassengers,
    required this.pricePerSeat,
    required this.status,
    required this.passengers,
    required this.createdAt,
    required this.updatedAt,
  });

  factory SharedRide.fromJson(Map<String, dynamic> json) {
    return SharedRide(
      id: json['id'] as int,
      driverId: json['driver_id'] as int,
      vehicleId: json['vehicle_id'] as int?,
      pickupLatitude: (json['pickup_latitude'] as num).toDouble(),
      pickupLongitude: (json['pickup_longitude'] as num).toDouble(),
      pickupAddress: json['pickup_address'] as String,
      dropoffLatitude: (json['dropoff_latitude'] as num).toDouble(),
      dropoffLongitude: (json['dropoff_longitude'] as num).toDouble(),
      dropoffAddress: json['dropoff_address'] as String,
      departureTime: DateTime.parse(json['departure_time'] as String),
      maxPassengers: json['max_passengers'] as int,
      currentPassengers: json['current_passengers'] as int,
      pricePerSeat: (json['price_per_seat'] as num).toDouble(),
      status: json['status'] as String,
      passengers: (json['passengers'] as List<dynamic>?)
              ?.map((p) => SharedRidePassenger.fromJson(p))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'driver_id': driverId,
      'vehicle_id': vehicleId,
      'pickup_latitude': pickupLatitude,
      'pickup_longitude': pickupLongitude,
      'pickup_address': pickupAddress,
      'dropoff_latitude': dropoffLatitude,
      'dropoff_longitude': dropoffLongitude,
      'dropoff_address': dropoffAddress,
      'departure_time': departureTime.toIso8601String(),
      'max_passengers': maxPassengers,
      'current_passengers': currentPassengers,
      'price_per_seat': pricePerSeat,
      'status': status,
      'passengers': passengers.map((p) => p.toJson()).toList(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  bool get hasAvailableSeats => currentPassengers < maxPassengers;
  bool get isFull => currentPassengers >= maxPassengers;

  @override
  List<Object?> get props => [
        id,
        driverId,
        vehicleId,
        pickupLatitude,
        pickupLongitude,
        pickupAddress,
        dropoffLatitude,
        dropoffLongitude,
        dropoffAddress,
        departureTime,
        maxPassengers,
        currentPassengers,
        pricePerSeat,
        status,
        passengers,
        createdAt,
        updatedAt,
      ];
}

/// Passenger in a shared ride
class SharedRidePassenger extends Equatable {
  final int id;
  final int passengerId;
  final String passengerName;
  final String? passengerPhotoUrl;
  final double pickupLatitude;
  final double pickupLongitude;
  final String pickupAddress;
  final double dropoffLatitude;
  final double dropoffLongitude;
  final String dropoffAddress;
  final String status; // 'pending', 'confirmed', 'picked_up', 'dropped_off', 'cancelled'
  final double price;
  final DateTime joinedAt;

  const SharedRidePassenger({
    required this.id,
    required this.passengerId,
    required this.passengerName,
    this.passengerPhotoUrl,
    required this.pickupLatitude,
    required this.pickupLongitude,
    required this.pickupAddress,
    required this.dropoffLatitude,
    required this.dropoffLongitude,
    required this.dropoffAddress,
    required this.status,
    required this.price,
    required this.joinedAt,
  });

  factory SharedRidePassenger.fromJson(Map<String, dynamic> json) {
    return SharedRidePassenger(
      id: json['id'] as int,
      passengerId: json['passenger_id'] as int,
      passengerName: json['passenger_name'] as String,
      passengerPhotoUrl: json['passenger_photo_url'] as String?,
      pickupLatitude: (json['pickup_latitude'] as num).toDouble(),
      pickupLongitude: (json['pickup_longitude'] as num).toDouble(),
      pickupAddress: json['pickup_address'] as String,
      dropoffLatitude: (json['dropoff_latitude'] as num).toDouble(),
      dropoffLongitude: (json['dropoff_longitude'] as num).toDouble(),
      dropoffAddress: json['dropoff_address'] as String,
      status: json['status'] as String,
      price: (json['price'] as num).toDouble(),
      joinedAt: DateTime.parse(json['joined_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'passenger_id': passengerId,
      'passenger_name': passengerName,
      'passenger_photo_url': passengerPhotoUrl,
      'pickup_latitude': pickupLatitude,
      'pickup_longitude': pickupLongitude,
      'pickup_address': pickupAddress,
      'dropoff_latitude': dropoffLatitude,
      'dropoff_longitude': dropoffLongitude,
      'dropoff_address': dropoffAddress,
      'status': status,
      'price': price,
      'joined_at': joinedAt.toIso8601String(),
    };
  }

  @override
  List<Object?> get props => [
        id,
        passengerId,
        passengerName,
        passengerPhotoUrl,
        pickupLatitude,
        pickupLongitude,
        pickupAddress,
        dropoffLatitude,
        dropoffLongitude,
        dropoffAddress,
        status,
        price,
        joinedAt,
      ];
}
